osbuild: share terminal formats between files

supakeen opened the pull request:
Terminal escape sequences for formatting were defined multiple times. This is a prelude to more extensive (prettier) formatting.

dvdhrm commented on osbuild/main_cli.py:
```diff
@@ -16,11 +16,7 @@
-BOLD = "\033[1m"
-RED = "\033[31m"
-GREEN = "\033[32m"
+from osbuild.markup import RESET, BOLD, RED, GREEN
```
Any reason to separate it from the other includes? Shouldn't it just go above `from osbuild.objectstore import ...`? Or maybe even above `osbuild.meta`, depending on how exactly we order it.

supakeen replied:
That fits our current style better, moved it up.

dvdhrm commented on osbuild/markup.py:
```diff
@@ -0,0 +1,6 @@
+RESET = "\033[0m"
+
+BOLD = "\033[1m"
+
+RED = "\033[31m"
+GREEN = "\033[32m"
```
I don't mind extracting them like this, but maybe as a followup we could group them better into a class like `vt`, that collects all vt-escape-sequences? Then we avoid importing each individual one everywhere and can just resort to `from markup import vt`?

supakeen replied:
I was anticipating the same follow up with the addition of an isatty() call to nullify the formatting if we're not outputting to a tty.

dvdhrm commented on osbuild/monitor.py:
```diff
@@ -18,9 +18,7 @@
-
-RESET = "\033[0m"
-BOLD = "\033[1m"
+from osbuild.markup import RESET, BOLD
```
Same here, drop the empty line between `import osbuild` and your import?

gicmo reviewed (changes requested):
Yeah, good idea, but I think we should put this into `utils` like we do with the rest of supporting code. I also wonder if we should make a general command line utils package and put it in there, maybe in a `vt` class like @dvdhrm suggested.

supakeen commented:
Moved this to `utils/`, added `__getattr__` to naively disable the formatting when stdout is not a tty. I know that doesn't catch all cases but it also doesn't make things worse :)

dvdhrm commented on osbuild/util/markup.py:
```diff
@@ -0,0 +1,28 @@
+
+
+    def __getattr__(self, name: str) -> str:
+        if not sys.stdout.isatty():
+            return ""
```
This runs an ioctl-syscall each time you fetch an escape-sequence, kinda working against the concept of IO buffering.

How about calling `isatty()` in the constructor of `VT`, and then creating either a VT object that supports escape sequences, or one that doesn't? This would also allow overriding the mode in the future, because if you pipe output into `less` or similar tools, `isatty()` returns false, even though you actually want coloring/etc.

Regardless, I think we can do that in a followup.

supakeen replied:
Agree, moved it to a property on the object. We can add overridable behaviour later.

dvdhrm commented on osbuild/main_cli.py:
```diff
@@ -14,15 +14,10 @@
 import osbuild
 import osbuild.meta
 import osbuild.monitor
+from osbuild.util.markup import vt
 from osbuild.objectstore import ObjectStore
```
```suggestion
import osbuild.monitor
from osbuild.objectstore import ObjectStore
from osbuild.util.markup import vt
```

`util` is ordered after `objectstore` in the alphabet, isn't it?

It feels like I am the only one obsessing with alphabetical include-ordering. Why doesn't this bother everyone? How can you live with yourself knowing the includes are unordered?

supakeen replied:
Haha, I'm used to autoformatting and I'm still internalizing the preferred import sorting here. Fixed.

dvdhrm reviewed (approved):
Two minor nits inline. Do with them whatever you like!

I like the way this is going! Feel free to merge as it is.

gicmo commented:
I still think that `markup` is very specific and it is more likely we gain more terminal related utilities so maybe `term.py` is better? Or maybe @dvdhrm has some other suggestion?

dvdhrm commented:
> I still think that `markup` is very specific and it is more likely we gain more terminal related utilities so maybe `term.py` is better? Or maybe @dvdhrm has some other suggestion?

True. term.py sounds fine to me. Technically these come from DEC-VT100 and friends, but `tty`, `terminal`, `term`, `vt`, all have been used interchangably in the past. So `term.py` sounds good to me.

supakeen commented:
Sounds good, I'll get the changes in soon :)

supakeen commented:
Merci for the review.

- Renamed to `term.py`.
- Instance name is now `fmt` from `term` (I don't like duplicating VT/term).
- `isatty` is now cached on the `VT` instance.

dvdhrm commented on osbuild/util/term.py:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    def __init__(self) -> None:
+        self.isatty = sys.stdout.isatty()
```
I think I would have done the `isatty` addition in a follow-up, so we first get the extraction into the module. But I don't mind either way. However, note that you now hard-code `stdout` in the module, while any code *using* this module might not necessarily know that it is specific to `stdout`. Just something to keep in mind. Not sure how to avoid this. One option would be to name the global instance to `stdout`/etc., but not sure about it.

supakeen replied:
I thought I mentioned this somewhere but yea it's hard to figure out. Perhaps removing the isatty check is 'just better'? :)

dvdhrm reviewed (approved):
I like it! One comment inline, but not sure this is worth delaying this.

dvdhrm commented:
`fmt` seems to be already taken as global name (standard library?)

gicmo commented:
> `fmt` seems to be already taken as global name (standard library?)

It is a local variable at: https://github.com/osbuild/osbuild/blob/a3bc5f7692d054d2457bffe3dde5f0ffb43ae8a2/osbuild/main_cli.py#L109

supakeen commented:
> `fmt` seems to be already taken as global name (standard library?)

It shouldn't collide with anything by default:

```
€ python3 -q
>>> fmt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'fmt' is not defined
```
